src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs

```diff
@@ -0,0 +1,523 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipes;
+using System.Linq;
+using System.Net.Sockets;
+using System.Runtime.InteropServices;
+using System.Security.Principal;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.NETCore.Client
+{
+    /// <summary>
+    /// An endpoint used to acquire the diagnostics stream for a runtime instance.
+    /// </summary>
+    internal interface IIpcEndpoint
+    {
+        /// <summary>
+        /// Wait for an available diagnostics connection to the runtime instance.
+        /// </summary>
+        /// <param name="token">The token to monitor for cancellation requests.</param>
+        /// <returns>
+        /// A task the completes when a diagnostics connection to the runtime instance becomes available.
+        /// </returns>
+        Task WaitForConnectionAsync(CancellationToken token);
+
+        /// <summary>
+        /// Connects to the underlying IPC Transport and opens a read/write-able Stream
+        /// </summary>
+        /// <returns>A Stream for writing and reading data to and from the target .NET process</returns>
+        Stream Connect();
+    }
+
+    internal abstract class BaseIpcEndpoint : IIpcEndpoint
+    {
+        // The amount of time to wait for a stream to be available for consumption by the Connect method.
+        // This should be a large timeout in order to allow the runtime instance to reconnect and provide
+        // a new stream once the previous stream has started its diagnostics request and response.
+        private static readonly TimeSpan ConnectWaitTimeout = TimeSpan.FromSeconds(30);
```

Comment:
It's more of an abundance of caution, not that I see any particular issue that was causing the reconnect to not happen fast enough. The feedback that I got was that we should have some timeout when waiting, even if it is large, instead of indefinitely waiting.

Reply:
I've reduced the timeout. It's still generous, but not overly so.